---
title: "Simulation Results"
output: html_notebook
---

```{r}
library(dplyr, warn.conflicts = FALSE)
library(tidyr)
library(ggplot2)
options(dplyr.summarise.inform = FALSE)
```

Read in data

```{r}
data.dir <- "../data"
pref <- c("afr_common_b2_eur_rare_b2",
          #"afr_common_b2_eur_unobs_b2",
          "afr_common_b5_eur_unobs_b5",
          "afr_common_h001_eur_rare_h001",
          #"afr_common_h001_eur_unobs_h001",
          "afr_rare_b2_eur_common_b2",
          "afr_rare_h001_eur_common_h001",
          #"afr_rare_b2_eur_unobs_b2",
          "afr_rare_b5_eur_unobs_b5",
          "afr_unobs_b9_eur_rare_b9",
          "same_freq_afr001_eur0015",
          "same_freq_afr002_eur001",
          "same_freq_afr_eur_v1")
type <- c("afr > 0.05 > eur > 0.01",
          "afr > 0.05 > eur ~ 0",
          "afr > 0.05 > eur > 0.01",
          "0.01 < afr < 0.05 < eur",
          "0.01 < afr < 0.05 < eur",
          "0.05 > afr > 0.01 > eur ~ 0",
          "0 ~ afr < 0.01 < eur < 0.05",
          "afr = eur > 0.05",
          "afr = eur > 0.05",
          "afr = eur > 0.05")
type.levels <- c("0 ~ afr < 0.01 < eur < 0.05",
                 "0.01 < afr < 0.05 < eur",
                 "afr = eur > 0.05",
                 "0.05 > afr > 0.01 > eur ~ 0",
                 "afr > 0.05 > eur > 0.01",
                 "afr > 0.05 > eur ~ 0")
eff <- c("afr = eur",
         "afr = eur",
         "afr < eur",
         "afr = eur",
         "afr > eur",
         "afr = eur",
         "afr = eur",
         "afr < eur",
         "afr > eur",
         "afr = eur")
eff.levels <- c("afr < eur",
                "afr = eur",
                "afr > eur")

names(type) <- pref
names(eff) <- pref

dat <- lapply(pref, function(p) {
    lapply(1:22, function(c) {
        f <- file.path(data.dir, paste0(p, "_effects_chr", c, ".rds"))
        readRDS(f) %>%
            mutate("relative_freq"=factor(type[p], levels=type.levels),
                   "effect_size"=factor(eff[p], levels=eff.levels))
    }) %>% bind_rows()
}) %>% bind_rows()
```


Calculate empirical power (fraction of p-values that are significant).

>  For the empirical power, you basically have a series of `n` bernouli random variables that take the value 1 if you find an association and 0 if you don't find an association, where the parameter `p` is the true power.  you estimate `p` (i.e. the empirical power you already calculated) by taking the average of all the 0s and 1s. The variance of the random variable that is that average is therefore `p(1-p)/n`, so the SE is `sqrt(p(1-p)/n)`. you can then use the usual formula of `p +/- qnorm(0.975)*SE` to get the upper and lower bounds of the 95% CI.

Construct 95% confidence intervals and see what fraction of observed values falls within them

```{r}
signif <- 5e-9

dat.stats <- dat %>%
    filter(!is.na(Est)) %>% # remove variants with freq=0 in one group
    mutate(beta_conf_lower=(Est + qnorm(0.025)*Est.SE),
           beta_conf_upper=(Est + qnorm(0.975)*Est.SE)) %>%
    group_by(relative_freq, effect_size, group) %>%
    summarise(n=n(), 
              emp_power=sum(Score.pval < signif)/n,
              beta_conf95=sum(beta > beta_conf_lower & beta < beta_conf_upper)/n,
              sum_sq_diff=sum((Est - beta)^2)) %>%
    mutate(power_SE=sqrt(emp_power*(1-emp_power)/n),
           power_conf_lower=(emp_power + qnorm(0.025)*power_SE),
           power_conf_upper=(emp_power + qnorm(0.975)*power_SE))

dat.stats

# ggplot(dat.stats, aes(group, emp_power, color=relative_freq, shape=effect_size,
#                       group=interaction(relative_freq, effect_size))) + 
#     geom_line() + geom_point() + scale_color_brewer(palette="Set1")
# 
# ggplot(dat.stats, aes(group, beta_conf95, color=relative_freq, shape=effect_size,
#                       group=interaction(relative_freq, effect_size))) + 
#     geom_line() + geom_point() + scale_color_brewer(palette="Set1")

p <- list()
for (b in unique(dat.stats$effect_size)) {
p[[b]] <- dat.stats %>%
        filter(effect_size %in% b) %>%
ggplot(aes(group, emp_power)) + 
    geom_col() +
    geom_errorbar(aes(ymin=power_conf_lower, ymax=power_conf_upper), width=0.4) +
    facet_wrap(~relative_freq, scales="free_y") +
    ylab("empirical power") +
    ggtitle(paste("effect size:", b, 
                  "\n panels = relative frequency"))
}

p
```


```{r}
# dat.signif <- dat %>%
#     group_by(relative_freq, effect_size, variant.id, group) %>%
#     summarise(is.signif=(!is.na(Score.pval) & min(Score.pval) < signif)) %>% 
#     pivot_wider(names_from=group, values_from=is.signif) %>%
#     summarise(none.signif=!(afr | eur | pooled),
#               indiv.only=((afr | eur) & !pooled),
#               pooled.only=(!(afr | eur) & pooled),
#               pooled.plus.one=(((afr & !eur) | (!afr & eur)) & pooled),
#               all.signif=(afr & eur & pooled)) %>% 
#     select(-variant.id) %>%
#     summarise(across(everything(), sum)) %>%
#     pivot_longer(none.signif:all.signif, names_to="result", values_to="n") %>%
#     mutate(result=factor(result, levels=c("none.signif", "indiv.only", "pooled.only",
#                                           "pooled.plus.one", "all.signif")))

dat.signif <- dat %>%
    group_by(relative_freq, effect_size, variant.id, group) %>%
    summarise(is.signif=(!is.na(Score.pval) & min(Score.pval) < signif)) %>% 
    pivot_wider(names_from=group, values_from=is.signif) %>%
    summarise(afr.only=(afr & !eur & !pooled),
              eur.only=(!afr & eur & !pooled),
              pooled.only=(!afr & !eur & pooled),
              pooled.plus.afr=(afr & !eur & pooled),
              pooled.plus.eur=(!afr & eur & pooled),
              all.signif=(afr & eur & pooled)) %>% 
    select(-variant.id) %>%
    summarise(across(everything(), sum)) %>%
    pivot_longer(afr.only:all.signif, names_to="result", values_to="n") %>%
    mutate(result=factor(result, levels=c("afr.only", "eur.only", "pooled.only",
                                          "pooled.plus.afr", "pooled.plus.eur",
                                          "all.signif"))) %>%
    group_by(relative_freq, effect_size) %>%
    mutate(frac_signif=n/sum(n))

dat.signif

p <- list()
for (b in unique(dat.signif$effect_size)) {
p[[b]] <- dat.signif %>%
        filter(effect_size %in% b) %>%
ggplot(aes(result, frac_signif)) + 
    geom_col() +
    facet_wrap(~relative_freq, scales="free_y") +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    ylab("fraction of signifiant variants") +
    ggtitle(paste("effect size:", b, 
                  "\n panels = relative frequency"))
}

p
```

```{r}
dat.pval <- dat %>%
    filter(!is.na(Score.pval)) %>%
    mutate(log10pval=-log10(Score.pval)) %>%
    select(relative_freq, effect_size, variant.id, group, log10pval) %>%
    pivot_wider(names_from=group, values_from=log10pval) %>%
    pivot_longer(all_of(c("afr", "eur")), names_to="group", values_to="indiv") %>%
    filter(!is.na(indiv))


p1 <- list()
for (b in unique(dat.pval$effect_size)) {
p1[[b]] <- dat.pval %>%
        filter(effect_size %in% b) %>%
ggplot(aes(indiv, pooled)) +
    geom_point() +
    geom_abline(slope=1, intercept=0, color="gray") +
    geom_vline(xintercept=-log10(signif), color="red", linetype=2) +
    geom_hline(yintercept=-log10(signif), color="red", linetype=2) +
    facet_grid(group~relative_freq, scales="free") +
    ggtitle(paste("effect size:", b, 
                  "\n columns = relative frequency",
                  "\n rows = strata"))
}

p1
```

```{r}
p.all <- list()
for (b in unique(dat.pval$effect_size)) {
    tmp.stats <- filter(dat.stats, effect_size %in% b)
    tmp.signif <- filter(dat.signif, effect_size %in% b)
    tmp.pval <- filter(dat.pval, effect_size %in% b)
    p.tmp <- list()
    for (f in unique(tmp.stats$relative_freq)) {
        p.tmp[[1]] <- tmp.stats %>%
            filter(relative_freq %in% f) %>%
            ggplot(aes(group, emp_power)) + 
            geom_col() +
            geom_errorbar(aes(ymin=power_conf_lower, ymax=power_conf_upper), width=0.4) +
            ylab("empirical power") +
            ggtitle(paste("effect size:", b, 
                          "\nrelative freq:", f))
        
        p.tmp[[2]] <- tmp.signif %>%
            filter(relative_freq %in% f) %>%
            ggplot(aes(result, frac_signif)) + 
            geom_col() +
            theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
            ylab("fraction of signifiant variants") +
            ggtitle(paste("effect size:", b, 
                          "\nrelative freq:", f))
        
        p.tmp[[3]] <- tmp.pval %>%
            filter(relative_freq %in% f) %>%
            ggplot(aes(indiv, pooled)) +
            geom_point() + coord_fixed() +
            geom_abline(slope=1, intercept=0, color="gray") +
            geom_vline(xintercept=-log10(signif), color="red", linetype=2) +
            geom_hline(yintercept=-log10(signif), color="red", linetype=2) +
            facet_wrap(~group) +
            ggtitle(paste("effect size:", b, 
                          "\nrelative frequency:", f))
        
        p.all[[paste(b,f)]] <- p.tmp
    }
}
```

```{r, fig.width=4, fig.height=4}
lapply(p.all, function(x) x[1:2])
```


```{r, fig.width=7, fig.height=4}
lapply(p.all, function(x) x[3])
```

# Estimated vs true beta

```{r}
i <- 1
p <- list()
for (b in unique(dat$effect_size)) {
    tmp <- filter(dat, effect_size %in% b)
    for (f in unique(tmp$relative_freq)) {
        tmp2 <- filter(tmp, relative_freq %in% f) 
        if (i <= 5) {
        p[[paste(b,f)]] <- tmp2 %>%
            ggplot(aes(beta, Est)) + 
            geom_boxplot() +
            facet_wrap(~group) +
            ggtitle(paste("effect size:", b, 
                          "\nrelative freq:", f))
        } else {
        p[[paste(b,f)]] <- tmp2 %>%
            ggplot(aes(beta, Est)) + 
            geom_point() +
            geom_abline(slope=1, intercept=0, color="gray") +
            facet_wrap(~group) +
            ggtitle(paste("effect size:", b, 
                          "\nrelative freq:", f))
            
        }
        i <- i + 1
    }
}

p
```

# Compute correlation with PCs

```{r}
# save all the variant.ids
var.id <- sort(dat$variant.id)
save(var.id, file="../data/simulation_variant_include_20210610.RData")
```

```{r}
library(gdsfmt)
files <- paste0("../data/sim_vars_20210610_pca_corr_chr", 1:22, ".gds")
corr <- do.call(rbind, lapply(unname(files), function(f) {
    c <- openfn.gds(f)
    tmp <- t(read.gdsn(index.gdsn(c, "correlation")))
    n_pcs <- ncol(tmp)
    colnames(tmp) <- paste0("PC", 1:n_pcs)
    tmp <- data.frame(tmp,
                      variant.id=readex.gdsn(index.gdsn(c, "snp.id")),
                      chr=readex.gdsn(index.gdsn(c, "chromosome")),
                      pos=readex.gdsn(index.gdsn(c, "position")),
                      stringsAsFactors=FALSE)
    closefn.gds(c)
    tmp
}))
saveRDS(corr, file="../data/sim_vars_20210610_pca_corr.rds")
```

```{r}
corr <- readRDS("../data/sim_vars_20210610_pca_corr.rds")
dat.corr <- dat %>%
    mutate(variant.id=as.integer(variant.id)) %>%
    left_join(corr) 

p <- list()
for (b in unique(dat.corr$effect_size)) {
    tmp <- filter(dat.corr, effect_size %in% b)
    for (f in unique(tmp$relative_freq)) {
        p[[paste(b,f)]] <- tmp %>%
            filter(relative_freq %in% f) %>%
        ggplot(aes(abs(PC1), Est/beta)) +
        geom_point() +
            geom_smooth(method="lm", color="gray") +
        facet_wrap(~group) +
            ggtitle(paste("effect size:", b, 
                          "\nrelative freq:", f))
    }
}

p
```

Meta-analysis

```{r}
meta <- function(beta, var) {
    W <- 1/var
    beta.fixed <- rowSums(beta*W, na.rm=T)/rowSums(W, na.rm=T)
    var.fixed <- 1/rowSums(W, na.rm=T)
    
    test.stat <- beta.fixed/sqrt(var.fixed)
    pval <- 2*pnorm(abs(test.stat), lower.tail = F)

    return(data.frame(beta = beta.fixed, 
                      var = var.fixed, 
                      test.stat = test.stat^2, 
                      pval = pval))
}

metal <- list()
for (b in unique(dat$effect_size)) {
    tmp <- filter(dat, effect_size %in% b)
    for (f in unique(tmp$relative_freq)) {
        tmp2 <- filter(tmp, relative_freq %in% f)
        tmp3 <- tmp2 %>%
            filter(group != "pooled") %>%
            select(variant.id, Est, Est.SE, group) %>%
            pivot_wider(names_from=group, values_from=c(Est, Est.SE))
        tmp.meta <- meta(beta=tmp3[,c("Est_afr", "Est_eur")], 
                         var=(tmp3[,c("Est.SE_afr", "Est.SE_eur")])^2)
        metal[[paste(b, f)]] <- cbind(tmp3, tmp.meta) %>%
            mutate(effect_size=b, relative_freq=f)
    }
}
meta.df <- bind_rows(metal) %>%
    mutate(group="meta") %>%
    mutate(Est.SE=sqrt(var)) %>%
    rename(Est=beta, Stat=test.stat) %>%
    select(-ends_with(c("afr", "eur")))

dat.meta <- dat %>%
    rename(Stat=Score.Stat, pval=Score.pval) %>%
    bind_rows(meta.df)
```


```{r}
signif <- 5e-9
p <- list()
for (b in unique(dat.meta$effect_size)) {
    tmp <- filter(dat.meta, effect_size %in% b)
    for (f in unique(tmp$relative_freq)) {
        tmp2 <- filter(tmp, relative_freq %in% f)
        
        p[[paste(b, f)]] <- tmp2 %>%
            filter(group %in% c("pooled", "meta")) %>%
            filter(!is.na(pval)) %>%
            mutate(log10pval=-log10(pval)) %>%
            select(variant.id, group, log10pval) %>%
            pivot_wider(names_from=group, values_from=log10pval) %>%
            ggplot(aes(meta, pooled)) +
            geom_point() + coord_fixed() +
            geom_abline(slope=1, intercept=0, color="gray") +
            geom_vline(xintercept=-log10(signif), color="red", linetype=2) +
            geom_hline(yintercept=-log10(signif), color="red", linetype=2) +
            ggtitle(paste("effect size:", b, 
                          "\nrelative frequency:", f))
    }
}

p
```