---
title: "variant selection"
output: html_notebook
---

```{r}
library(dplyr)
library(ggplot2)
#remotes::install_github("UW-GAC/simulate_phenotypes/simphen", upgrade=FALSE)
remotes::install_local("/sbgenomics/workspace/simulate_phenotypes/simphen", upgrade=FALSE)
library(simphen)
#devtools::load_all("/sbgenomics/workspace/simulate_phenotypes/simphen")
```

Frequency of variants in chromosome 1 in AfrAm and EurAm strata

```{r}
afr <- readRDS("/sbgenomics/project-files/freeze8_AfrAm_allele_freq_chr1.rds")
eur <- readRDS("/sbgenomics/project-files/freeze8_EurAm_allele_freq_chr1.rds")

all.equal(afr$variant.id, eur$variant.id)

freq <- rename(afr, freq.afr=alt.freq) %>%
  bind_cols(select(eur, freq.eur=alt.freq)) %>%
  filter(freq.afr > 0 | freq.eur > 0) # remove monomorphic
nrow(freq)

rm(afr, eur) # save memory

saveRDS(freq, file="/sbgenomics/output-files/freeze8_AfrAm_EurAm_allele_freq_chr1.rds")
```

Select observed variants

```{r}
obs_both <- var_multiple_strata(freq$freq.afr, freq$freq.eur)
length(obs_both)
obs_afr <- var_single_stratum(freq$freq.afr, freq$freq.eur, min.freq=0.1, max.freq=0.001)
obs_eur <- var_single_stratum(freq$freq.eur, freq$freq.afr, min.freq=0.1, max.freq=0.001)
length(obs_afr)
length(obs_eur)
```

```{r}
obs_afr <- var_single_stratum(freq$freq.afr, freq$freq.eur, min.freq=0.01, max.freq=0.001)
obs_eur <- var_single_stratum(freq$freq.eur, freq$freq.afr, min.freq=0.01, max.freq=0.001)
length(obs_afr)
length(obs_eur)
```

```{r}
obs_afr <- var_single_stratum(freq$freq.afr, freq$freq.eur, min.freq=0.001, max.freq=0.0001)
obs_eur <- var_single_stratum(freq$freq.eur, freq$freq.afr, min.freq=0.001, max.freq=0.0001)
length(obs_afr)
length(obs_eur)
```


Load GDS file to get genotypes

```{r}
library(SeqArray)
gds <- seqOpen("/sbgenomics/project-files/freeze8_gds/freeze.8.pass_poly.gtonly.minDP0_chr1.gds")

samples.afr <- readRDS("/sbgenomics/project-files/strata/freeze8_samples_AfrAm_2020-07-23.rds")
samples.eur <- readRDS("/sbgenomics/project-files/strata/freeze8_samples_EurAm_2020-07-23.rds")
seqSetFilter(gds, sample.id=c(samples.afr, samples.eur))
```

Get index of variants for each variant.id (required for variant_assoc function)

```{r}
variant.id <- seqGetData(gds, "variant.id")
freq <- left_join(freq, data.frame(variant.id, variant.sel=1:length(variant.id)))
```

What values of h2 give us reasonable power in both populations?

```{r}
for (h2 in c(0.005, 0.003, 0.002, 0.0015, 0.001)) {
  print(paste("h2=", h2, "Afr power:", power(N=length(samples.afr), h2=h2)))
  print(paste("h2=", h2, "Eur power:", power(N=length(samples.eur), h2=h2)))
  print(paste("h2=", h2, "Total power:", power(N=length(c(samples.afr, samples.eur)), h2=h2)))
}
```


# Same allele freq; same effect size and heritability

```{r}
same_freq <- freq %>%
  filter(freq.afr < 0.5, freq.eur < 0.5,
         freq.afr > 0.05, freq.eur > 0.05, 
         abs(freq.afr - freq.eur) < 0.001)
nrow(same_freq)
```

sample outcome and match order to samples in GDS file

```{r}
outcomes <- readRDS("/sbgenomics/project-files/sim_outcomes_AfrEur_varComp_3_7_named.rds")
this.outcome <- outcomes[[sample(length(outcomes), 1)]]
```

get PCs and kinship

```{r}
load("/sbgenomics/project-files/pcair_results.RData")
pcs <- as.data.frame(pca$vectors)
names(pcs) <- paste0("PC", 1:ncol(pcs))
pcs$sample.id <- rownames(pcs)

load("/sbgenomics/project-files/covmat_blocks_AfrEur_2020-07-24.RData")

dat <- data.frame(outcome=this.outcome, sample.id=names(this.outcome), 
                  stringsAsFactors=FALSE)
dat <- dplyr::left_join(dat, pcs)
dat <- Biobase::AnnotatedDataFrame(dat)
```

get variant effect - test with subset of samples

```{r}
geno <- variant_genotypes(gds, variant.sel=same_freq$variant.sel[1])
strata <- list(afr=samples.afr[1:1000], eur=samples.eur[1:1000])
v1 <- variant_assoc(geno, h2=list(afr=0.001, eur=0.001), varComp=c(3,7),
                    dat=dat, outcome="outcome", cov.mat=covmat, 
                    strata=strata, covars=paste0("PC", 1:11))
v1
```

Use LD pruning to select independent variants. Prune first in pooled sample with MAF threshold 0.01, then in each population separately with no threshold.

```{r}
seqSetFilter(gds, verbose=FALSE)
strata <- list(afr=samples.afr, eur=samples.eur)

library(SNPRelate)

set.seed(7)
pruned.all <- snpgdsLDpruning(gds, sample.id=unlist(strata), snp.id=same_freq$variant.id,
                              maf=0.01, method="corr", 
                              slide.max.bp=10e6, ld.threshold=sqrt(0.1))
pruned.all <- unlist(pruned.all, use.names=FALSE)

set.seed(71)
pruned.afr <- snpgdsLDpruning(gds, sample.id=samples.afr, snp.id=pruned.all,
                              remove.monosnp=FALSE, method="corr",  
                              slide.max.bp=10e6, ld.threshold=sqrt(0.1))
pruned.afr <- unlist(pruned.afr, use.names=FALSE)

set.seed(72)
pruned.eur <- snpgdsLDpruning(gds, sample.id=samples.eur, snp.id=pruned.afr,
                              remove.monosnp=FALSE, method="corr",  
                              slide.max.bp=10e6, ld.threshold=sqrt(0.1))
pruned <- unlist(pruned.eur, use.names=FALSE)
length(pruned)

saveRDS(pruned, file="/sbgenomics/output-files/freeze8_same_freq_pruned_afr_eur_v1.rds")
```

Add effects of first 10 pruned genotypes to outcome

```{r}
pruned <- readRDS("/sbgenomics/output-files/freeze8_same_freq_pruned_afr_eur_v1.rds")
geno <- variant_genotypes(gds, variant.id=pruned[1:10])
v2 <- variant_assoc(geno, h2=list(afr=0.001, eur=0.001), varComp=c(3,7),
                    dat=dat, outcome="outcome", cov.mat=covmat, 
                    strata=strata, covars=paste0("PC", 1:11))
v2
saveRDS(v2, file="/sbgenomics/output-files/samefreq_assoc_test_10pruned_2020-11-13.rds")
```

Test some variants individually and compare

```{r}
v2.1.indiv <- variant_assoc(geno[,1,drop=FALSE], h2=list(afr=0.001, eur=0.001), varComp=c(3,7),
                    dat=dat, outcome="outcome", cov.mat=covmat, 
                    strata=strata, covars=paste0("PC", 1:11))
v2.5.indiv <- variant_assoc(geno[,5,drop=FALSE], h2=list(afr=0.001, eur=0.001), varComp=c(3,7),
                    dat=dat, outcome="outcome", cov.mat=covmat, 
                    strata=strata, covars=paste0("PC", 1:11))
v2.10.indiv <- variant_assoc(geno[,10,drop=FALSE], h2=list(afr=0.001, eur=0.001), varComp=c(3,7),
                    dat=dat, outcome="outcome", cov.mat=covmat, 
                    strata=strata, covars=paste0("PC", 1:11))

v2.indiv <- as.data.frame(data.table::rbindlist(list(v2.1.indiv, v2.5.indiv, v2.10.indiv)))
saveRDS(v2.indiv, file="/sbgenomics/output-files/samefreq_assoc_test_10pruned_indiv_2020-11-13.rds")
```

```{r}
chk <- inner_join(select(v2, variant.id, group, Est.comb=Est),
           select(v2.indiv, variant.id, group, Est.indiv=Est)) %>%
  mutate(diff=Est.comb-Est.indiv)
chk
saveRDS(chk, file="/sbgenomics/output-files/samefreq_assoc_test_10pruned_indiv_comb_diff_2020-11-13.rds")
```

Save some files to use in scripts

```{r}
saveRDS(strata, file="/sbgenomics/output-files/freeze8_strata_afr_eur.rds")
dat$outcome <- NULL
saveRDS(dat, file="/sbgenomics/output-files/freeze8_AnnotatedDataFrame_afr_eur_pcs.rds")
```


Repeat combining afr and eur freq for all chromosomes

```{r}
for (c in 2:22) {
  afr <- readRDS(paste0("/sbgenomics/project-files/freeze8_AfrAm_allele_freq_chr", c, ".rds"))
  eur <- readRDS(paste0("/sbgenomics/project-files/freeze8_EurAm_allele_freq_chr", c, ".rds"))

  stopifnot(all.equal(afr$variant.id, eur$variant.id))

  freq <- rename(afr, freq.afr=alt.freq) %>%
    bind_cols(select(eur, freq.eur=alt.freq)) %>%
    filter(freq.afr > 0 | freq.eur > 0) # remove monomorphic
  print(nrow(freq))

  rm(afr, eur) # save memory

  saveRDS(freq, file=paste0("/sbgenomics/output-files/freeze8_AfrAm_EurAm_allele_freq_chr", c, ".rds"))
}
```


Repeat selecting variants with same allele freq; same effect size and heritability for all chromosomes

```{r}
for (c in 1:22) { # include chr1 here since we only saved the pruned list before
  if (c == 1) {
    f <- paste0("/sbgenomics/project-files/freeze8_AfrAm_EurAm_allele_freq_chr", c, ".rds")
  } else {
    f <- paste0("/sbgenomics/output-files/freeze8_AfrAm_EurAm_allele_freq_chr", c, ".rds")
  }
  freq <- readRDS(f)
  same_freq <- freq %>%
    filter(freq.afr < 0.5, freq.eur < 0.5,
           freq.afr > 0.05, freq.eur > 0.05, 
           abs(freq.afr - freq.eur) < 0.001)
  print(nrow(same_freq))
  saveRDS(same_freq, file=paste0("/sbgenomics/output-files/same_freq_afr_eur_v1_chr", c, ".rds"))
}
```



# Select variants with different frequencies in each population

```{r}
for (c in 1:22) {
  freq <- readRDS(paste0("/sbgenomics/project-files/freeze8_allele_freq/freeze8_AfrAm_EurAm_allele_freq_chr", c, ".rds"))

  # Common in Afr, rare in Eur
  sel <- freq %>%
    filter(freq.afr < 0.5, freq.eur < 0.05,
           freq.afr > 0.05, freq.eur > 0.01, 
           abs(freq.afr - freq.eur) > 0.05)
  print(nrow(sel))
  saveRDS(sel, file=paste0("/sbgenomics/output-files/afr_common_eur_rare_v1_chr", c, ".rds"))
  
  # Common in Eur, rare in Afr
  sel <- freq %>%
    filter(freq.afr < 0.05, freq.eur < 0.5,
           freq.afr > 0.01, freq.eur > 0.05, 
           abs(freq.afr - freq.eur) > 0.05)
  print(nrow(sel))
  saveRDS(sel, file=paste0("/sbgenomics/output-files/afr_rare_eur_common_v1_chr", c, ".rds"))
  
  # Common in Afr, unobs in Eur
  sel <- freq %>%
    filter(freq.afr < 0.5, freq.eur < 0.0005,
           freq.afr > 0.05,
           abs(freq.afr - freq.eur) > 0.05)
  print(nrow(sel))
  saveRDS(sel, file=paste0("/sbgenomics/output-files/afr_common_eur_unobs_v1_chr", c, ".rds"))
  
  # Common in Eur, unobs in Afr
  sel <- freq %>%
    filter(freq.afr < 0.0005, freq.eur < 0.5,
           freq.eur > 0.05,
           abs(freq.afr - freq.eur) > 0.05)
  print(nrow(sel))
  saveRDS(sel, file=paste0("/sbgenomics/output-files/afr_unobs_eur_common_v1_chr", c, ".rds"))

}
```

No variants (technically only one) in the "common eur, unobs afr category". Delete these files with nothing in them.

```{r}
for (c in 1:22) {
  unlink(paste0("/sbgenomics/output-files/afr_unobs_eur_common_v1_chr", c, ".rds"))
}
```

Save only variant.id in rds file

```{r}
for (c in 1:22) {
  # Common in Afr, rare in Eur
  sel <- readRDS(paste0("/sbgenomics/project-files/afr_common_eur_rare_v1_chr", c, ".rds"))
  var <- sel$variant.id
  saveRDS(var, file=paste0("/sbgenomics/output-files/afr_common_eur_rare_v1_variant_id_chr", c, ".rds"))
  
  # Common in Eur, rare in Afr
  sel <- readRDS(paste0("/sbgenomics/project-files/afr_rare_eur_common_v1_chr", c, ".rds"))
  var <- sel$variant.id
  saveRDS(var, file=paste0("/sbgenomics/output-files/afr_rare_eur_common_v1_variant_id_chr", c, ".rds"))
  
  # Common in Afr, unobs in Eur
  sel <- readRDS(paste0("/sbgenomics/project-files/afr_common_eur_unobs_v1_chr", c, ".rds"))
  var <- sel$variant.id
  saveRDS(var, file=paste0("/sbgenomics/output-files/afr_common_eur_unobs_v1_variant_id_chr", c, ".rds"))
}
```


# Select variants rare in afr and unobserved in eur

```{r}
for (c in 1:22) {
  freq <- readRDS(paste0("/sbgenomics/project-files/freeze8_allele_freq/freeze8_AfrAm_EurAm_allele_freq_chr", c, ".rds"))

  # Rare in Afr, unobs in Eur
  sel <- freq %>%
    filter(freq.afr < 0.05, freq.eur < 0.0005,
           freq.afr > 0.01)
  print(nrow(sel))
  saveRDS(sel, file=paste0("/sbgenomics/output-files/afr_rare_eur_unobs_v1_chr", c, ".rds"))
  
  var <- sel$variant.id
  saveRDS(var, file=paste0("/sbgenomics/output-files/afr_rare_eur_unobs_v1_variant_id_chr", c, ".rds"))
}
```


# Select variants unobserved in afr and rare in eur

```{r}
for (c in 1:22) {
  freq <- readRDS(paste0("/sbgenomics/project-files/freeze8_allele_freq/freeze8_AfrAm_EurAm_allele_freq_chr", c, ".rds"))

  # Rare in Afr, unobs in Eur
  sel <- freq %>%
    filter(freq.eur < 0.05, freq.afr < 0.0005,
           freq.eur > 0.01)
  print(nrow(sel))
  saveRDS(sel, file=paste0("/sbgenomics/output-files/afr_unobs_eur_rare_v1_chr", c, ".rds"))
  
  var <- sel$variant.id
  saveRDS(var, file=paste0("/sbgenomics/output-files/afr_unobs_eur_rare_v1_variant_id_chr", c, ".rds"))
}
```


What beta values should we use when the variants are rare?

```{r}
afr_rare <- lapply(1:22, function(c) {
  readRDS(paste0("/sbgenomics/project-files/afr_rare_eur_unobs_v1_chr", c, ".rds"))
}) %>% bind_rows()
```

```{r}
(p <- median(afr_rare$freq.afr))
# assume HWE to get variance of genotype
varG <- 2*p*(1-p)
varComp <- c(3,7)
h2 <- 0.001
htot2 <- (varComp[1] + h2*varComp[2])/sum(varComp)
beta <- sqrt((h2*varComp[2])/((1-htot2)*varG))
beta
```

higher power in afr:

```{r}
h2 <- 0.002
htot2 <- (varComp[1] + h2*varComp[2])/sum(varComp)
beta <- sqrt((h2*varComp[2])/((1-htot2)*varG))
beta
```

Can we make a table of beta, h2, freq, and power for each group?

```{r}
strata <- readRDS("../data/freeze8_strata_afr_eur.rds")
N <- lapply(strata, length)
N$pooled <- N$afr + N$eur
```

```{r}
beta <- seq(0.2, 0.7, 0.1)
freq <- c(0.005, 0.01, 0.02, 0.05, 0.1)
varComp <- c(3,7)
x <- lapply(c("afr", "eur", "pooled"), function(group) {
    lapply(freq, function(p) {
        lapply(beta, function(b) {
            varG <- 2*p*(1-p)
            h2 <- (varG*b^2) / (varG*b^2 + sum(varComp))
            pow <- power(N[[group]], h2)
            tibble(group, freq=p, beta=b, h2=h2, power=pow)
        }) %>% bind_rows()
    }) %>% bind_rows()
}) %>% bind_rows()
write.csv(x, file="../power_stats.txt", quote=F, row.names=F)
```
