---
title: "variant selection"
output: html_notebook
---

```{r}
library(dplyr)
library(ggplot2)
#remotes::install_github("UW-GAC/simulate_phenotypes/simphen", upgrade=FALSE)
remotes::install_local("/sbgenomics/workspace/simulate_phenotypes/simphen", upgrade=FALSE)
library(simphen)
#library(devtools)
#load_all("/sbgenomics/workspace/simulate_phenotypes/simphen")
```

Frequency of variants in chromosome 1 in AfrAm and EurAm strata

```{r}
afr <- readRDS("/sbgenomics/project-files/freeze8_allele_freq_chr1_AfrAm.rds")
eur <- readRDS("/sbgenomics/project-files/freeze8_allele_freq_chr1_EurAm.rds")

all.equal(afr$variant.id, eur$variant.id)

freq <- rename(afr, freq.afr=alt.freq) %>%
  bind_cols(select(eur, freq.eur=alt.freq)) %>%
  filter(freq.afr > 0 | freq.eur > 0) # remove monomorphic
nrow(freq)

rm(afr, eur) # save memory
```

Select observed variants

```{r}
obs_both <- var_multiple_strata(freq$freq.afr, freq$freq.eur)
length(obs_both)
obs_afr <- var_single_stratum(freq$freq.afr, freq$freq.eur, min.freq=0.1, max.freq=0.001)
obs_eur <- var_single_stratum(freq$freq.eur, freq$freq.afr, min.freq=0.1, max.freq=0.001)
length(obs_afr)
length(obs_eur)
```

```{r}
obs_afr <- var_single_stratum(freq$freq.afr, freq$freq.eur, min.freq=0.01, max.freq=0.001)
obs_eur <- var_single_stratum(freq$freq.eur, freq$freq.afr, min.freq=0.01, max.freq=0.001)
length(obs_afr)
length(obs_eur)
```

```{r}
obs_afr <- var_single_stratum(freq$freq.afr, freq$freq.eur, min.freq=0.001, max.freq=0.0001)
obs_eur <- var_single_stratum(freq$freq.eur, freq$freq.afr, min.freq=0.001, max.freq=0.0001)
length(obs_afr)
length(obs_eur)
```


Generate outcome from covariance matrix

```{r}
covmat <- GWASTools::getobj("/sbgenomics/project-files/pcrelate_kinshipMatrix_sparseDeg4_v2.RData")
# subset to only samples we are using
samples.afr <- readRDS("/sbgenomics/project-files/strata/freeze8_samples_AfrAm_2020-07-23.rds")
samples.eur <- readRDS("/sbgenomics/project-files/strata/freeze8_samples_EurAm_2020-07-23.rds")
ind <- which(rownames(covmat) %in% c(samples.afr, samples.eur))
covmat <- covmat[ind,ind]
```

First we need to find the block indices for the matrix - just do this once for large matrices (needs between 8 and 30 GB of memory to calculate indices)

```{r}
blocks <- simphen:::block_indices(covmat)
```

Save matrix and blocks together

```{r, eval=FALSE}
length(blocks)
table(sapply(blocks, length))
save(covmat, blocks, file="/sbgenomics/workspace/covmat_blocks_AfrEur_2020-07-24.RData")
```

See how long it takes to compute outcome

```{r}
load("/sbgenomics/workspace/covmat_blocks_AfrEur_2020-07-24.RData")
```

```{r}
varComp <- c(3,7)
system.time(outcome <- simphen:::outcome_from_covMat_block_indices(covmat, blocks, varComp))
```


Load GDS file to get genotypes

```{r}
library(SeqArray)
gds <- seqOpen("/sbgenomics/project-files/freeze8_gds/freeze.8.pass_poly.gtonly.minDP0_chr1.gds")
seqSetFilter(gds, sample.id=c(samples.afr, samples.eur))
```
