<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Stephanie M. Gogarten" />

<meta name="date" content="2020-07-10" />

<title>Simulating phenotypes</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Simulating phenotypes</h1>
<h4 class="author">Stephanie M. Gogarten</h4>
<h4 class="date">2020-07-10</h4>



<p>For this example we will use 1000 genomes data. We start with a covariance matrix calcuated using the GENESIS package. We found pairwise kinship with <code>pcrelate</code> and created a block diagonal matrix of kinship using <code>pcrelateToMatrix</code>. We generate a random normal outcome adjusted for covariance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(simphen)
<span class="kw">data</span>(<span class="st">&quot;1KG_pcrelate_blockDiag_Matrix&quot;</span>)
varComp &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">40</span>, <span class="dv">100</span>)
<span class="kw">set.seed</span>(<span class="dv">8</span>)
outcome &lt;-<span class="st"> </span><span class="kw">outcome_from_covMat_blocks</span>(blockDiagMatrix1KG, varComp)</code></pre></div>
<p>Next we define strata using 1000 genomes super populations, and calculate variant frequency in each stratum.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SeqArray)
gdsfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;1KG_phase3_chr1_SNVsubset.gds&quot;</span>, <span class="dt">package=</span><span class="st">&quot;simphen&quot;</span>)
gds &lt;-<span class="st"> </span><span class="kw">seqOpen</span>(gdsfile)
sample.id &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;sample.id&quot;</span>)
pops &lt;-<span class="st"> </span><span class="kw">seqGetData</span>(gds, <span class="st">&quot;sample.annotation/Population&quot;</span>)
super.pops &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;EAS&quot;</span>=<span class="kw">c</span>(<span class="st">&quot;CHB&quot;</span>, <span class="st">&quot;JPT&quot;</span>, <span class="st">&quot;CHS&quot;</span>, <span class="st">&quot;CDX&quot;</span>, <span class="st">&quot;KHV&quot;</span>),
                   <span class="st">&quot;EUR&quot;</span>=<span class="kw">c</span>(<span class="st">&quot;CEU&quot;</span>, <span class="st">&quot;TSI&quot;</span>, <span class="st">&quot;FIN&quot;</span>, <span class="st">&quot;GBR&quot;</span>, <span class="st">&quot;IBS&quot;</span>),
                   <span class="st">&quot;AFR&quot;</span>=<span class="kw">c</span>(<span class="st">&quot;YRI&quot;</span>, <span class="st">&quot;LWK&quot;</span>, <span class="st">&quot;GWD&quot;</span>, <span class="st">&quot;MSL&quot;</span>, <span class="st">&quot;ESN&quot;</span>, <span class="st">&quot;ASW&quot;</span>, <span class="st">&quot;ACB&quot;</span>),
                   <span class="st">&quot;AMR&quot;</span>=<span class="kw">c</span>(<span class="st">&quot;MXL&quot;</span>, <span class="st">&quot;PUR&quot;</span>, <span class="st">&quot;CLM&quot;</span>, <span class="st">&quot;PEL&quot;</span>),
                   <span class="st">&quot;SAS&quot;</span>=<span class="kw">c</span>(<span class="st">&quot;GIH&quot;</span>, <span class="st">&quot;PJL&quot;</span>, <span class="st">&quot;BEB&quot;</span>, <span class="st">&quot;STU&quot;</span>, <span class="st">&quot;ITU&quot;</span>))
strata &lt;-<span class="st"> </span><span class="kw">lapply</span>(super.pops, <span class="cf">function</span>(x) sample.id[pops <span class="op">%in%</span><span class="st"> </span>x])

<span class="co"># alternate allele frequency</span>
freq &lt;-<span class="st"> </span><span class="kw">do.call</span>(cbind, <span class="kw">lapply</span>(strata, <span class="cf">function</span>(x) {
    <span class="kw">seqSetFilter</span>(gds, <span class="dt">sample.id=</span>x, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
    <span class="kw">seqAlleleFreq</span>(gds, <span class="dt">ref.allele=</span><span class="dv">1</span>)
}))
<span class="kw">seqResetFilter</span>(gds, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</code></pre></div>
<p>Select variants based on frequency in different strata</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select variants in AFR but not EUR</span>
var_afr &lt;-<span class="st"> </span><span class="kw">var_single_stratum</span>(freq[,<span class="st">&quot;AFR&quot;</span>], freq[,<span class="st">&quot;EUR&quot;</span>])

<span class="co"># select variants in both</span>
var_both &lt;-<span class="st"> </span><span class="kw">var_multiple_strata</span>(freq[,<span class="st">&quot;AFR&quot;</span>], freq[,<span class="st">&quot;EUR&quot;</span>])</code></pre></div>
<p>Add an effect for a variant in only one stratum</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqSetFilter</span>(gds, <span class="dt">variant.sel=</span>var_afr[<span class="dv">1</span>], <span class="dt">verbose=</span><span class="ot">FALSE</span>)
geno &lt;-<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">seqGetData</span>(gds, <span class="st">&quot;$dosage_alt&quot;</span>))
outcome.eff &lt;-<span class="st"> </span>outcome <span class="op">+</span><span class="st"> </span><span class="kw">variant_effect</span>(geno, <span class="dt">h2=</span><span class="fl">0.01</span>, <span class="dt">varComp=</span>varComp)</code></pre></div>
<p>Test association</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Biobase)
<span class="kw">library</span>(SeqVarTools)
<span class="kw">library</span>(GENESIS)
dat &lt;-<span class="st"> </span><span class="kw">AnnotatedDataFrame</span>(<span class="kw">data.frame</span>(sample.id, <span class="dt">outcome=</span>outcome.eff, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>))
nullmod &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(dat, <span class="dt">outcome=</span><span class="st">&quot;outcome&quot;</span>, <span class="dt">cov.mat=</span>blockDiagMatrix1KG, 
                        <span class="dt">verbose=</span><span class="ot">FALSE</span>)
seqData &lt;-<span class="st"> </span><span class="kw">SeqVarData</span>(gds, <span class="dt">sampleData=</span>dat)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarBlockIterator</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestSingle</span>(iterator, nullmod, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc</code></pre></div>
<pre><code>##   variant.id chr     pos allele.index n.obs       freq MAC    Score Score.SE
## 1     100334   1 3240623            1  2504 0.03813898 191 6.837323 1.173246
##   Score.Stat   Score.pval      Est    Est.SE        PVE
## 1   5.827696 5.619798e-09 4.967154 0.8523359 0.01356843</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># test with only AFR samples</span>
<span class="kw">resetIterator</span>(iterator, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
nullmod_afr &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(dat, <span class="dt">outcome=</span><span class="st">&quot;outcome&quot;</span>, <span class="dt">cov.mat=</span>blockDiagMatrix1KG,
                            <span class="dt">sample.id=</span>strata[[<span class="st">&quot;AFR&quot;</span>]], <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc_afr &lt;-<span class="st"> </span><span class="kw">assocTestSingle</span>(iterator, nullmod_afr, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc_afr</code></pre></div>
<pre><code>##   variant.id chr     pos allele.index n.obs      freq MAC    Score Score.SE
## 1     100334   1 3240623            1   661 0.1361573 180 5.370229 1.016257
##   Score.Stat   Score.pval      Est    Est.SE        PVE
## 1   5.284322 1.261712e-07 5.199789 0.9840031 0.04211351</code></pre>
<p>Add an effect common in both strata</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqSetFilter</span>(gds, <span class="dt">variant.sel=</span>var_both[<span class="dv">1</span>], <span class="dt">verbose=</span><span class="ot">FALSE</span>)
geno &lt;-<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">seqGetData</span>(gds, <span class="st">&quot;$dosage_alt&quot;</span>))
outcome.eff &lt;-<span class="st"> </span>outcome <span class="op">+</span><span class="st"> </span><span class="kw">variant_effect</span>(geno, <span class="dt">h2=</span><span class="fl">0.01</span>, <span class="dt">varComp=</span>varComp)</code></pre></div>
<p>Add an effect different across strata</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">afr &lt;-<span class="st"> </span>sample.id <span class="op">%in%</span><span class="st"> </span>strata[[<span class="st">&quot;AFR&quot;</span>]]
outcome.eff.afr &lt;-<span class="st"> </span>outcome[afr] <span class="op">+</span><span class="st"> </span><span class="kw">variant_effect</span>(geno[afr], <span class="dt">h2=</span><span class="fl">0.01</span>, <span class="dt">varComp=</span>varComp)
eur &lt;-<span class="st"> </span>sample.id <span class="op">%in%</span><span class="st"> </span>strata[[<span class="st">&quot;EUR&quot;</span>]]
outcome.eff.eur &lt;-<span class="st"> </span>outcome[eur] <span class="op">+</span><span class="st"> </span><span class="kw">variant_effect</span>(geno[eur], <span class="dt">h2=</span><span class="fl">0.05</span>, <span class="dt">varComp=</span>varComp)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat &lt;-<span class="st"> </span><span class="kw">AnnotatedDataFrame</span>(<span class="kw">data.frame</span>(sample.id, <span class="dt">outcome=</span><span class="ot">NA</span>, <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>))
dat<span class="op">$</span>outcome[afr] &lt;-<span class="st"> </span>outcome.eff.afr
dat<span class="op">$</span>outcome[eur] &lt;-<span class="st"> </span>outcome.eff.eur
nullmod &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(dat, <span class="dt">outcome=</span><span class="st">&quot;outcome&quot;</span>, <span class="dt">cov.mat=</span>blockDiagMatrix1KG, 
                        <span class="dt">verbose=</span><span class="ot">FALSE</span>)
iterator &lt;-<span class="st"> </span><span class="kw">SeqVarBlockIterator</span>(seqData, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc &lt;-<span class="st"> </span><span class="kw">assocTestSingle</span>(iterator, nullmod, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc</code></pre></div>
<pre><code>##   variant.id chr     pos allele.index n.obs      freq MAC   Score Score.SE
## 1     100019   1 3232378            1  1164 0.6391753 840 9.03746 2.146517
##   Score.Stat   Score.pval      Est    Est.SE        PVE
## 1   4.210291 2.550417e-05 1.961453 0.4658711 0.01524209</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">resetIterator</span>(iterator, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
nullmod_afr &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(dat, <span class="dt">outcome=</span><span class="st">&quot;outcome&quot;</span>, <span class="dt">cov.mat=</span>blockDiagMatrix1KG,
                            <span class="dt">sample.id=</span>strata[[<span class="st">&quot;AFR&quot;</span>]], <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc_afr &lt;-<span class="st"> </span><span class="kw">assocTestSingle</span>(iterator, nullmod_afr, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc_afr</code></pre></div>
<pre><code>##   variant.id chr     pos allele.index n.obs      freq MAC    Score Score.SE
## 1     100019   1 3232378            1   661 0.8441755 206 1.535462 1.114388
##   Score.Stat Score.pval      Est    Est.SE         PVE
## 1   1.377852  0.1682489 1.236421 0.8973536 0.002876475</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">resetIterator</span>(iterator, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
nullmod_eur &lt;-<span class="st"> </span><span class="kw">fitNullModel</span>(dat, <span class="dt">outcome=</span><span class="st">&quot;outcome&quot;</span>, <span class="dt">cov.mat=</span>blockDiagMatrix1KG,
                            <span class="dt">sample.id=</span>strata[[<span class="st">&quot;EUR&quot;</span>]], <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc_eur &lt;-<span class="st"> </span><span class="kw">assocTestSingle</span>(iterator, nullmod_eur, <span class="dt">verbose=</span><span class="ot">FALSE</span>)
assoc_eur</code></pre></div>
<pre><code>##   variant.id chr     pos allele.index n.obs      freq MAC    Score Score.SE
## 1     100019   1 3232378            1   503 0.3697813 372 4.760229 1.273329
##   Score.Stat   Score.pval      Est    Est.SE        PVE
## 1   3.738413 0.0001851859 2.935936 0.7853431 0.02784009</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqClose</span>(gds)</code></pre></div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
