---
title: "Simulating phenotypes"
author: "Stephanie M. Gogarten"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Simulating phenotypes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```

```{r}
library(simphen)
library(Biobase)
library(SeqArray)
library(SeqVarTools)
library(GENESIS)
library(SNPRelate)
```

# Simulate outcome from covariance

For this example we will use 1000 genomes data. We start with a covariance matrix calcuated using the GENESIS package. We found pairwise kinship with `pcrelate` and created a block diagonal matrix of kinship using `pcrelateToMatrix`. We generate a random normal outcome adjusted for covariance.

```{r}
data("1KG_pcrelate_blockDiag_Matrix")
varComp <- c(4, 10)
set.seed(8)
outcome <- outcome_from_covMat_blocks(blockDiagMatrix1KG, varComp)
head(outcome)
```

# Define strata

Next we define strata using 1000 genomes super populations, and calculate variant frequency in each stratum.

```{r}
data("1KG_phase3_annot")
super.pops <- unique(annot$Super.Population)
strata <- lapply(super.pops, function(x) annot$sample.id[annot$Super.Population %in% x])
names(strata) <- super.pops

gdsfile <- system.file("extdata", "1KG_phase3_chr1_SNVsubset.gds", package="simphen")
gds <- seqOpen(gdsfile)

# alternate allele frequency
freq <- do.call(cbind, lapply(strata, function(x) {
    seqSetFilter(gds, sample.id=x, verbose=FALSE)
    seqAlleleFreq(gds, ref.allele=1)
}))

seqResetFilter(gds, verbose=FALSE)
```

```{r}
# select variants in AFR but not EUR
var_afr <- var_single_stratum(freq[,"AFR"], freq[,"EUR"])

# select variants in both
var_both <- var_multiple_strata(freq[,"AFR"], freq[,"EUR"])
```


# Adding one variant at a time

## Add an effect for a variant in only one stratum

Since the covariance matrix is block diagonal, the order of samples is
not the same as in the GDS file. Re-order the outcome vector to match
the genotypes.

```{r}
sample.id <- seqGetData(gds, "sample.id")
outcome.gdsord <- outcome[sample.id]
head(outcome.gdsord)
```

```{r}
seqSetFilter(gds, variant.sel=var_afr[1], verbose=FALSE)
geno <- as.vector(seqGetData(gds, "$dosage_alt"))
eff <- variant_effect(G=geno, h2=0.01, varComp=varComp)
eff$beta
outcome.eff <- outcome.gdsord + eff$Gbeta
```

Test association

```{r}
dat <- annot
dat$outcome <- outcome.eff
nullmod <- fitNullModel(dat, outcome="outcome", cov.mat=blockDiagMatrix1KG, 
                        covars="Population", verbose=FALSE)
seqData <- SeqVarData(gds, sampleData=dat)
iterator <- SeqVarBlockIterator(seqData, verbose=FALSE)
assoc <- assocTestSingle(iterator, nullmod, verbose=FALSE)
assoc
```

```{r}
# test with only AFR samples
resetIterator(iterator, verbose=FALSE)
nullmod_afr <- fitNullModel(dat, outcome="outcome", cov.mat=blockDiagMatrix1KG,
                            covars="Population", sample.id=strata[["AFR"]], verbose=FALSE)
assoc_afr <- assocTestSingle(iterator, nullmod_afr, verbose=FALSE)
assoc_afr
```

Convenience function to add effect by strata and test in one step. We will consider only two groups for simplicity.

```{r}
dat$outcome <- outcome.gdsord
geno <- variant_genotypes(gds, variant.sel=var_afr[1])
strata2 <- strata[c("AFR", "EUR")]
beta <- list(AFR=1.33, EUR=1.33)
variant_assoc(geno, beta=beta, varComp=varComp, dat=dat, outcome="outcome", 
              cov.mat=blockDiagMatrix1KG, strata=strata2, covars="Population")
```


## Add an effect common in both strata - specify effect size

```{r}
geno <- variant_genotypes(gds, variant.sel=var_both[4])
beta <- list(AFR=1.3, EUR=1.3)
variant_assoc(geno, beta=beta, varComp=varComp, dat=dat, outcome="outcome", 
              cov.mat=blockDiagMatrix1KG, strata=strata2, covars="Population")
```


## Add an effect different across strata

```{r}
geno <- variant_genotypes(gds, variant.sel=var_both[4])
beta <- list(AFR=1.5, EUR=1.3)
variant_assoc(geno, beta=beta, varComp=varComp, dat=dat, outcome="outcome", 
              cov.mat=blockDiagMatrix1KG, strata=strata2, covars="Population")
```



# Add multiple variant effects to same outcome

## Use LD pruning to select independent variants

Prune first in pooled sample with MAF threshold 0.01, then in each population separately with no threshold.

```{r}
seqSetFilter(gds, verbose=FALSE)
# use a high ld.threshold only because the variant set is so small
pruned <- ldprune_strata(gds, strata2, maf=0.01, ld.threshold=0.5,
                         seed=list(all=7, AFR=71, EUR=72))
```

Add effects from first 5 pruned variants

```{r}
geno <- variant_genotypes(gds, variant.id=pruned[1:5], sample.id=unlist(strata2))
beta <- list(AFR=1.2, EUR=1.1)
variant_assoc(geno, beta=beta, varComp=varComp, dat=dat, outcome="outcome", 
              cov.mat=blockDiagMatrix1KG, strata=strata2, covars="Population")
```

```{r}
seqClose(gds)
```
