---
title: "Simulating phenotypes"
author: "Stephanie M. Gogarten"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Simulating phenotypes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```

# Simulate outcome from covariance

For this example we will use 1000 genomes data. We start with a covariance matrix calcuated using the GENESIS package. We found pairwise kinship with `pcrelate` and created a block diagonal matrix of kinship using `pcrelateToMatrix`. We generate a random normal outcome adjusted for covariance.

```{r}
library(simphen)
data("1KG_pcrelate_blockDiag_Matrix")
varComp <- c(4, 10)
set.seed(8)
outcome <- outcome_from_covMat_blocks(blockDiagMatrix1KG, varComp)
head(outcome)
```

## Define strata

Next we define strata using 1000 genomes super populations, and calculate variant frequency in each stratum.

```{r}
library(SeqArray)
gdsfile <- system.file("extdata", "1KG_phase3_chr1_SNVsubset.gds", package="simphen")
gds <- seqOpen(gdsfile)
sample.id <- seqGetData(gds, "sample.id")
pops <- seqGetData(gds, "sample.annotation/Population")
super.pops <- list("EAS"=c("CHB", "JPT", "CHS", "CDX", "KHV"),
                   "EUR"=c("CEU", "TSI", "FIN", "GBR", "IBS"),
                   "AFR"=c("YRI", "LWK", "GWD", "MSL", "ESN", "ASW", "ACB"),
                   "AMR"=c("MXL", "PUR", "CLM", "PEL"),
                   "SAS"=c("GIH", "PJL", "BEB", "STU", "ITU"))
strata <- lapply(super.pops, function(x) sample.id[pops %in% x])

# alternate allele frequency
freq <- do.call(cbind, lapply(strata, function(x) {
    seqSetFilter(gds, sample.id=x, verbose=FALSE)
    seqAlleleFreq(gds, ref.allele=1)
}))
seqResetFilter(gds, verbose=FALSE)
```


```{r}
# select variants in AFR but not EUR
var_afr <- var_single_stratum(freq[,"AFR"], freq[,"EUR"])

# select variants in both
var_both <- var_multiple_strata(freq[,"AFR"], freq[,"EUR"])
```


# Adding one variant at a time

## Add an effect for a variant in only one stratum

Since the covariance matrix is block diagonal, the order of samples is
not the same as in the GDS file. Re-order the outcome vector to match
the genotypes.

```{r}
outcome.gdsord <- outcome[sample.id]
head(outcome.gdsord)
```



```{r}
seqSetFilter(gds, variant.sel=var_afr[1], verbose=FALSE)
geno <- as.vector(seqGetData(gds, "$dosage_alt"))
eff <- variant_effect(G=geno, h2=0.01, varComp=varComp)
eff$beta
outcome.eff <- outcome.gdsord + eff$Gbeta
```

Test association

```{r}
library(Biobase)
library(SeqVarTools)
library(GENESIS)
dat <- AnnotatedDataFrame(data.frame(sample.id, outcome=outcome.eff, stringsAsFactors=FALSE))
nullmod <- fitNullModel(dat, outcome="outcome", cov.mat=blockDiagMatrix1KG, 
                        verbose=FALSE)
seqData <- SeqVarData(gds, sampleData=dat)
iterator <- SeqVarBlockIterator(seqData, verbose=FALSE)
assoc <- assocTestSingle(iterator, nullmod, verbose=FALSE)
assoc
```

```{r}
# test with only AFR samples
resetIterator(iterator, verbose=FALSE)
nullmod_afr <- fitNullModel(dat, outcome="outcome", cov.mat=blockDiagMatrix1KG,
                            sample.id=strata[["AFR"]], verbose=FALSE)
assoc_afr <- assocTestSingle(iterator, nullmod_afr, verbose=FALSE)
assoc_afr
```

Convenience function to add effect and test in one step.

```{r}
dat$outcome <- outcome.gdsord
variant_assoc(variant.sel=var_afr[1], beta=1.3, varComp=varComp, 
              gdsobj=gds, dat=dat, outcome="outcome", 
              cov.mat=blockDiagMatrix1KG)
```

```{r}
dat.afr <- dat[dat$sample.id %in% strata[["AFR"]],]
variant_assoc(variant.sel=var_afr[1], beta=1.3, varComp=varComp, 
              gdsobj=gds, dat=dat.afr, outcome="outcome", 
              cov.mat=blockDiagMatrix1KG)
```

## Add an effect common in both strata - specify effect size

```{r}
variant_assoc(variant.sel=var_both[4], beta=1.5, varComp=varComp, 
              gdsobj=gds, dat=dat, outcome="outcome", 
              cov.mat=blockDiagMatrix1KG)
```


```{r}
# test with only AFR samples
variant_assoc(variant.sel=var_both[4], beta=1.5, varComp=varComp, 
              gdsobj=gds, dat=dat.afr, outcome="outcome", 
              cov.mat=blockDiagMatrix1KG)
```

## Add an effect different across strata

```{r}
seqSetFilter(gds, variant.sel=var_both[4], verbose=FALSE)
geno <- as.vector(seqGetData(gds, "$dosage_alt"))

afr <- sample.id %in% strata[["AFR"]]
eff.afr <- variant_effect(geno[afr], beta=1.2, varComp=varComp)
power(N=sum(afr), h2=eff.afr$h2) 
outcome.eff.afr <- outcome.gdsord[afr] + eff.afr$Gbeta

eur <- sample.id %in% strata[["EUR"]]
eff.eur <- variant_effect(geno[eur], beta=1.5, varComp=varComp)
power(N=sum(eur), h2=eff.eur$h2) 
outcome.eff.eur <- outcome.gdsord[eur] + eff.eur$Gbeta
```

```{r}
dat <- AnnotatedDataFrame(data.frame(sample.id, outcome=NA, stringsAsFactors=FALSE))
dat$outcome[afr] <- outcome.eff.afr
dat$outcome[eur] <- outcome.eff.eur
nullmod <- fitNullModel(dat, outcome="outcome", cov.mat=blockDiagMatrix1KG, 
                        verbose=FALSE)

seqData <- SeqVarData(gds, sampleData=dat)
iterator <- SeqVarBlockIterator(seqData, verbose=FALSE)
assoc <- assocTestSingle(iterator, nullmod, verbose=FALSE)
assoc
```

```{r}
resetIterator(iterator, verbose=FALSE)
nullmod_afr <- fitNullModel(dat, outcome="outcome", cov.mat=blockDiagMatrix1KG,
                            sample.id=strata[["AFR"]], verbose=FALSE)
assoc_afr <- assocTestSingle(iterator, nullmod_afr, verbose=FALSE)
assoc_afr
```

```{r}
resetIterator(iterator, verbose=FALSE)
nullmod_eur <- fitNullModel(dat, outcome="outcome", cov.mat=blockDiagMatrix1KG,
                            sample.id=strata[["EUR"]], verbose=FALSE)
assoc_eur <- assocTestSingle(iterator, nullmod_eur, verbose=FALSE)
assoc_eur
```


# Use LD pruning to select independent variants

Prune first in pooled sample with MAF threshold 0.01, then in each population separately with no threshold.

```{r}
library(SNPRelate)
seqSetFilter(gds, verbose=FALSE)
# use a high ld.threshold only because the variant set is so small
pruned.all <- snpgdsLDpruning(gds, sample.id=c(strata[["AFR"]], strata[["EUR"]]),
                              maf=0.01, method="corr", ld.threshold=0.5)
pruned.all <- unlist(pruned.all, use.names=FALSE)
pruned.afr <- snpgdsLDpruning(gds, sample.id=strata[["AFR"]], snp.id=pruned.all,
                              remove.monosnp=FALSE, method="corr", ld.threshold=0.5)
pruned.afr <- unlist(pruned.afr, use.names=FALSE)
pruned.eur <- snpgdsLDpruning(gds, sample.id=strata[["EUR"]], snp.id=pruned.afr,
                              remove.monosnp=FALSE, method="corr", ld.threshold=0.5)
pruned <- unlist(pruned.eur, use.names=FALSE)
length(pruned)
```



```{r}
seqClose(gds)
```
